src/com/ka/spreadsheet/diff/UnifiedDiffSpreadSheetDiffCallback.java

```diff
@@ -0,0 +1,172 @@
+package com.ka.spreadsheet.diff;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.ka.spreadsheet.diff.SpreadSheetUtils.CELL_INTERNAL_TO_USER;
+
+// Output format, inspired by the traditional Unified Diff format (but using cell references instead of line numbers):
+//      +--- column 1
+//      V
+//        --- file_1_name!sheet_1_name
+//        +++ file_2_name!sheet_1_name
+//        @@ cell_1_ref,cell_n_ref cell_1_ref,cell_n_ref @@
+//        -old_cell_1_data
+//        -...
+//        -old_cell_n_data
+//        +new_cell_1_data
+//        +...
+//        +new_cell_n_data
+//        @@@ ...
+//        ...
+//        --- file_1_name!sheet_2_name
+//        +++ file_2_name!sheet_2_name
+//        ...
+// Each cell data line is always present, even if the data is empty (thus just "-" or "+).
+public class UnifiedDiffSpreadSheetDiffCallback implements SpreadSheetDiffCallback {
+
+  private final String lineSeparator = System.getProperty("line.separator");
+  private List<DiffCell> diffCells = new ArrayList<DiffCell>();
+
+  // This formatter requires that each cell in the pair of workbooks will be processed in worksheet-row-column order.
+  // SpreadSheetDiffer.java does that, but it's not a contract.  If that changes some day, set this to true to resort
```

Comment:
So the code is fine. But one issue I have with this approach is that it's keeping all the diff cells in memory. Which is likely to cause OOMs.

I think you may as well assume that the diff will be in worksheet-row-column order AND add an assertion for that assumption by keeping a running counter for max_row_seen, max_column_seen.

Then you can afford to only keep at most one contiguous chunk of diffs (max one whole row) in memory. Does that work?

Reply:
Yup, that makes sense.  I was trying not to add a constraint to the Differ API, and I'd much rather stream the output than buffer it.  I'll take a crack at reorganizing the code along these lines.

Reply:
That made the unified diff code simpler.  I documented the differ/callback contract, and implemented enforcement in both callback modules.